### PR TITLE
Faster & better Fuse.Profiling

### PR DESCRIPTION
**Pull request by Duckers**
Optimizations and improvements to make Fuse.Profiling more precise and usfeul

### CODE REVIEW

File: Source/Fuse.Common/Profiling.uno [hunk 1]
```diff
@@ -334,21 +245,20 @@ namespace Fuse
 			Write(id);
 		}
 
+		static string[] _cachedStrings = new string[0xFF];
```
Comment (kusma) [inline Source/Fuse.Common/Profiling.uno]: This should probably be `0x100` instead.

File: Source/Fuse.Common/Profiling.uno [hunk 2]
```diff
@@ -334,21 +245,20 @@ namespace Fuse
...
-			Write(Command.CacheString);
-			Write(id);
-			Write(str);
+			var hash = (byte)(str.GetHashCode() & 0xFF);
+			if (_cachedStrings[hash] != str)
+			{
+				Write(Command.CacheString);
+				Write(hash);
+				Write(str);
+				_cachedStrings[hash] = str;
```
Comment (kusma) [inline Source/Fuse.Common/Profiling.uno]: This will fail in the case where `str.GetHashCode() & 0xFF == 0xFF`, since _cachedStrings is 255 elements big, as opposed to 256.

Comment (Duckers) [inline Source/Fuse.Common/Profiling.uno]: Nice catch :) fixed now

File: Source/Fuse.Elements/Element.Drawing.uno [hunk 3]
```diff
@@ -66,21 +66,17 @@ namespace Fuse.Elements
...
 			}
 
-			var visibleRect = GetVisibleViewportInvertPixelRect(dc, RenderBoundsWithEffects);
-			if (visibleRect.Size.X == 0 || visibleRect.Size.Y == 0)
-			{
-				if defined(FUSELIBS_PROFILING)
-					Fuse.Profiling.EndRegion(Uno.Diagnostics.Clock.GetSeconds() - t);
-				return;
-			}
-
```
Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: This chance seems a bit dubious to me...  reducing profiling overhead have the flip-side of us missing the time spent inside `GetVisibleViewportInvertPixelRect` from our profiling...

Comment (Duckers) [inline Source/Fuse.Elements/Element.Drawing.uno]: Not missing it, it will be added to the parent's time block.

Comment (Duckers) [inline Source/Fuse.Elements/Element.Drawing.uno]: Whether viewport culling is something the parent or child does is a philosophical question.

Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: Yeah, exactly. It's being accounted wrong.

Comment (Duckers) [inline Source/Fuse.Elements/Element.Drawing.uno]: Having thousands of nodes in the profiling report that take insignificant amount of time is less useful than just compounding them to the parent's time block, imho.

Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: Reporting thousands of nodes culling-time on their parent seems like a much worse option to me.

Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: And more importantly, optimizing the profiler at the cost of it's accuracy seems like a bad trade-off.

Comment (Duckers) [inline Source/Fuse.Elements/Element.Drawing.uno]: This is improving its accuracy by reducing the overhead of the profiler.

Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: It's decreasing the accuracy by accounting information on the wrong node. The data you get our is garbage at this point. It doesn't matter if collecting it was fast.

Comment (Duckers) [inline Source/Fuse.Elements/Element.Drawing.uno]: I disagree. This improves the usefulness of the profiler. As I said, whether viewport culling is an activity of the parent or child is a philosophical question.

Asking for some pragmatism here. I'm not going to revert this.

Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: If you consider doing something like reporting the cull-time as an event if it took a lot of time, you might get the best of both worlds, by the way. As is, this is IMO an move to the worse. Which is sad, because it's holding back otherwise very nice changes. How about you split this part out into a separate PR?

Comment (Duckers) [inline Source/Fuse.Elements/Element.Drawing.uno]: The old implementation makes the profiler borderline useless in the very common use case of having a long list of nodes where most of the nodes are outside the viewport. The profiler has a not insignificant overhead per node of data it needs to report back home.

Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: I think you have no idea how expensive reading `RenderBoundsWithEffects` can be (worst case is full subtree traversal, ). This is very useful information, and you don't want to get that reported on the wrong element. That makes your data tell you the wrong story.

I already suggested an alternative approach that avoids the bulk of the profiling-cost and allows you to catch the cases where this ends up taking a significant amount of time.

Another alternative is to add profiling inside `CalcRenderBoundsWithEffects()`, which will only trigger when the cache invalid.

**Comment by Duckers** [comment]
@kusma : added reporting of time spent in culling as separate event on parent

**Comment by kusma** [comment]
@Duckers: That's a great compromise, thanks!

File: Source/Fuse.Elements/Element.Drawing.uno [hunk 4]
```diff
@@ -58,6 +58,9 @@ namespace Fuse.Elements
 			return fb;
 		}
 
+		extern (FUSELIBS_PROFILING) double _childCullTime;
+		extern (FUSELIBS_PROFILING) double cullTime; // to avoid local variable unused warning
```
Comment (kusma) [inline Source/Fuse.Elements/Element.Drawing.uno]: You don't need this. You can do `extern double cullTime;` instead.
